src/Microsoft.Health.Dicom.Api/Web/LazyMultipartReadOnlyStream.cs

```diff
@@ -0,0 +1,228 @@
+            {
+#pragma warning disable CA1835 // Prefer the 'Memory'-based overloads for 'ReadAsync' and 'WriteAsync', (Azure.Storage.LazyLoadingReadOnlyStream does not implement that.)
+                _bufferLength = await _currentStreamContent.Stream.ReadAsync(_buffer, 0, _bufferSize, cancellationToken);
+#pragma warning restore CA1835 // Prefer the 'Memory'-based overloads for 'ReadAsync' and 'WriteAsync'
+                _bufferPosition = 0;
+            }
+
+            // We will return the minimum of remainingBytesInBuffer and the count provided by the user
+            int remainingBytesInBuffer = _bufferLength - _bufferPosition;
+            int bytesToWrite = Math.Min(remainingBytesInBuffer, count);
+
+            // copy
+            Array.Copy(_buffer, _bufferPosition, buffer, offset, bytesToWrite);
+            _bufferPosition += bytesToWrite;
+
+            return bytesToWrite;
+        }
+
+        private int CopyStringToBuffer(string multipartHeader)
+        {
+            byte[] startBoundary = EncodeStringToByteArray(multipartHeader);
+            int bytesToWrite = startBoundary.Length;
+            Array.Copy(startBoundary, 0, _buffer, 0, bytesToWrite);
+            _bufferPosition = 0;
+            _bufferLength = bytesToWrite;
+
+            return bytesToWrite;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+            // Return the buffer to the pool if we're called from Dispose or a finalizer
+            if (_buffer != null)
+            {
+                ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
+                _buffer = null;
+            }
+        }
+
+        #region MultiPart boundary write
+        private string GetHeaderBoundary(bool isStart)
+        {
+            var multiPartStringBuilder = new StringBuilder();
+            if (isStart)
+            {
+                multiPartStringBuilder.Append(GetStartBoundary());
+
+            }
+            else
+            {
+                multiPartStringBuilder.Append(GetIntermediateBoundary());
+            }
+            multiPartStringBuilder.Append(GetContentHeader(_currentStreamContent.Headers));
+            return multiPartStringBuilder.ToString();
+        }
+
+        private byte[] EncodeStringToByteArray(string input)
+        {
+            return _defaultHttpEncoding.GetBytes(input);
+        }
+
+        private string GetStartBoundary()
+        {
+            return "--" + _boundary + Crlf;
+        }
+
+        private string GetIntermediateBoundary()
+        {
+            return Crlf + "--" + _boundary + Crlf;
+        }
+
+        private string GetTerminatingBoundary()
+        {
+            return Crlf + "--" + _boundary + "--" + Crlf;
+        }
+
+        private static string GetContentHeader(IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers)
+        {
+            StringBuilder builder = new StringBuilder();
```

Comment:
Should we use a new StringBuilder here or just use the one in the calling method?